File: algorithms/sorting/quick_sort.py
# Quick Sort (divide and conquer)

not_sorted_array = [2, 8, 5, 3, 9, 4, 1]
print("Before: ", not_sorted_array)

# Time Complexity:      Worst: θ(n ** 2), AVG: O(n log n), Best: Ω(n log n)
# Space Complexity :    Worst: O(n), Best: Ω(log n)
def partition(array, low, high):
    pivot = array[low]
    leftwall = low + 1

    for i in range(low + 1, high + 1):
        if array[i] < pivot:
            array[i], array[leftwall] = array[leftwall], array[i]
            leftwall += 1

    array[low], array[leftwall - 1] = array[leftwall - 1], array[low]

    return leftwall - 1

def quick_sort(array, low, high):
    if low < high:
        pivot_location = partition(array, low, high)
        quick_sort(array, low, pivot_location - 1)
        quick_sort(array, pivot_location + 1, high)

    return array


print("After:  ", quick_sort(not_sorted_array.copy(), 0, len(not_sorted_array) - 1))


# More pythonic style
def quick_sort_pythonic(array):
    if len(array) <= 1:
        return array

    pivot = array[0]
    left = [x for x in array[1:] if x < pivot]
    right = [x for x in array[1:] if x >= pivot]

    return quick_sort_pythonic(left) + [pivot] + quick_sort_pythonic(right)

print("After (Pythonic):  ", quick_sort_pythonic(not_sorted_array.copy()))

# Idea:
# Мы берём элемент массива — pivot (опорный элемент) — и раскладываем все числа относительно него:
#
# слева → всё, что меньше pivot
# справа → всё, что больше pivot
#
# После этого сам pivot уже оказывается на своём правильном месте в отсортированном массиве.
#
# Теперь мы делаем то же самое:
#
# рекурсивно сортируем левую часть
# рекурсивно сортируем правую часть
#
# И всё.

# Intuition:
# Представь, что pivot — это “разделитель мира”:
#
# Он не просто сравнивает элементы, он фиксирует свою финальную позицию сразу
#
# И дальше задача разбивается на две меньшие независимые задачи.

File: cpp_algorithms/sorting/quick_sort.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

// Quick Sort (Divide and Conquer)
//
// Time Complexity:
// Worst: O(n**2)
// Average: O(n log n)
// Best: O(n log n)
//
// Space Complexity:
// Worst: O(n)
// Average/Best: O(log n)

int partition(vector<int>& array, int low, int high) {
    int pivot = array[low];
    int leftwall = low + 1;

    for (int i = low + 1; i <= high; ++i) {
        if (array[i] < pivot) {
            swap(array[i], array[leftwall]);
            leftwall++;
        }
    }

    swap(array[low], array[leftwall - 1]);

    return leftwall - 1;
}

void quick_sort(vector<int>& array, int low, int high) {
    if (low < high) {
        int pivot_location = partition(array, low, high);

        quick_sort(array, low, pivot_location - 1);
        quick_sort(array, pivot_location + 1, high);
    }
}

int main() {
    vector<int> array = {2, 8, 5, 3, 9, 4, 1};

    cout << "Before: ";
    for (int x : array) {
        cout << x << ' ';
    }
    cout << '\n';

    quick_sort(array, 0, array.size() - 1);

    cout << "After:  ";
    for (int x : array) {
        cout << x << ' ';
    }
    cout << '\n';

    return 0;
}
